### PR TITLE
ILM fix the init step to actually be retryable

### PR DESCRIPTION
**Pull request by andreidan**
We marked the `init` ILM step as retryable but our test used `waitUntil`
without an assert so we didn’t catch the fact that we were not actually
able to retry this step as our ILM state didn’t contain any information
about the policy execution (as we were in the process of initialising
it).

This commit manually sets the current step to `init` when we’re moving
the ilm policy into the ERROR step (this enables us to successfully
move to the error step and later retry the step)

### CODE REVIEW

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 1]
```diff
@@ -1120,26 +1120,26 @@ public void testRolloverStepRetriesUntilRolledOverIndexIsDeleted() throws Except
         // {@link org.elasticsearch.xpack.core.ilm.ErrorStep} in order to retry the failing step. As {@link #assertBusy}
         // increases the wait time between calls exponentially, we might miss the window where the policy is on
         // {@link WaitForRolloverReadyStep} and the move to `attempt-rollover` request will not be successful.
-        waitUntil(() -> {
+        assertThat(waitUntil(() -> {
```
Comment (probakowski) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: I'd consider using `assertTrue`, we have long body here, it's hard to tell what we asserting here. The same applies to all instances below

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java [hunk 2]
```diff
@@ -47,13 +47,17 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
...
+            try {
+                long parsedOriginationDate = parseIndexNameAndExtractDate(index.getName());
+                indexMetadataBuilder.settingsVersion(indexMetaData.getSettingsVersion() + 1)
+                    .settings(Settings.builder()
+                        .put(indexMetaData.getSettings())
+                        .put(LifecycleSettings.LIFECYCLE_ORIGINATION_DATE, parsedOriginationDate)
+                        .build()
+                    );
+            } catch (Exception e) {
+                throw new InitializePolicyException(e.getMessage(), e);
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: What was the reason for wrapping this in an `ElasticsearchException` type exception?

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: It could be changed to just `assertBusy(() -> {...}, 30, TimeUnit.SECONDS)` also and use assertions to signal success/failure

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java [hunk 3]
```diff
@@ -1120,26 +1120,26 @@ public void testRolloverStepRetriesUntilRolledOverIndexIsDeleted() throws Except
...
             } catch (IOException e) {
                 return false;
             }
-        }, 30, TimeUnit.SECONDS);
+        }, 30, TimeUnit.SECONDS), is(true));
 
         // Similar to above, using {@link #waitUntil} as we want to make sure the `attempt-rollover` step started failing and is being
         // retried (which means ILM moves back and forth between the `attempt-rollover` step and the `error` step)
-        waitUntil(() -> {
+        assertThat("ILM did not start retrying the attempt-rollover step", waitUntil(() -> {
```
Comment (dakrone) [inline x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: I think the `assertThat` -> `waitUntil` -> `is(true)` is a little hard to follow, I think just a single `assertBusy()` would be better, since you can add the assertions in the body itself rather than returning a boolean?

**Review by dakrone — COMMENTED** [review]
I left a couple of comments about this, in particular I'm not sure we need to add a new exception type just for this?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java (hunk 2 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: I'm not entirely happy with using an exception for flow control but since we don't have any lifecycle state yet (as we're just initialising the policy) I used this exception to signal the current failing step is the "init policy step" when moving the policy into the `ERROR` step.

Another option would be to pass the current step key to `IndexLifecycleTransition.moveClusterStateToErrorStep` but this seemed more intrusive.

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 1 above)
Comment (andreidan) [reply x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: I've changed the `assertThat(waitUntil(), is(true))` to `assertTrue(waitUntil())` but the reason to use `waitUntil` as opposed to `assertBusy` is to avoid having this test flake. We are asserting that the failed step and the retry count are bound to particular values. The fact that we want to check the failed step is what we expect it to be means we have to *catch* ILM in the ERROR step, but with retryable steps ILM keeps moving back and forth between the failing step (on retry) and the ERROR step (when the step execution fails).

`assertBusy` increases the wait time exponentially once it gets to 1 second, while `waitUntil` keeps it at 1 second. This means we're bound to catch ILM in the ERROR step using `waitUntil` (but with assertBusy we probe the ILM state only at seconds 1, 2, 4, 8 and 16)

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/ilm/qa/multi-node/src/test/java/org/elasticsearch/xpack/ilm/TimeSeriesLifecycleActionsIT.java]: Ahh okay, I see the difference between them now, in that case I agree that `assertTrue` would be better than `assertThat` in this case

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java [hunk 4]
```diff
@@ -47,13 +47,17 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
...
         IndexMetaData.Builder indexMetadataBuilder = IndexMetaData.builder(indexMetaData);
         if (shouldParseIndexName(indexMetaData.getSettings())) {
-            long parsedOriginationDate = parseIndexNameAndExtractDate(index.getName());
-            indexMetadataBuilder.settingsVersion(indexMetaData.getSettingsVersion() + 1)
-                .settings(Settings.builder()
-                    .put(indexMetaData.getSettings())
-                    .put(LifecycleSettings.LIFECYCLE_ORIGINATION_DATE, parsedOriginationDate)
-                    .build()
-                );
+            try {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: I think we may want to move the `try` to surround more of the function (for example, the `fromIndexMetadata(...)` call

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyException.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+package org.elasticsearch.xpack.core.ilm;
+
+import org.elasticsearch.ElasticsearchException;
+
+/**
+ * Exception thrown when a problem is encountered while initialising an ILM policy for an index.
+ */
+public class InitializePolicyException extends ElasticsearchException {
+
+    public InitializePolicyException(String msg, Throwable cause, Object... args) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyException.java]: Can we include the policy name in the exception somewhere? I think that might be helpful if the setting were to change but the error was still there

File: x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java [hunk 6]
```diff
@@ -133,8 +134,13 @@ static ClusterState moveClusterStateToErrorStep(Index index, ClusterState cluste
         ElasticsearchException.generateThrowableXContent(causeXContentBuilder, STACKTRACE_PARAMS, cause);
         causeXContentBuilder.endObject();
         LifecycleExecutionState currentState = LifecycleExecutionState.fromIndexMetadata(idxMeta);
-        Step.StepKey currentStep = Objects.requireNonNull(LifecycleExecutionState.getCurrentStepKey(currentState),
-            "unable to move to an error step where there is no current step, state: " + currentState);
+        Step.StepKey currentStep;
+        if (cause instanceof InitializePolicyException) {
```
Comment (dakrone) [inline x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java]: Can you add a comment about why we check for this particular exception here?

**Review by dakrone — CHANGES_REQUESTED** [review]
I left a few more comments, thanks Andrei!

**Comment by andreidan** [comment]
@elasticmachine update branch

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java [hunk 7]
```diff
@@ -25,8 +27,11 @@
     public static final StepKey KEY = new StepKey(INITIALIZATION_PHASE, "init", "init");
     private static final Logger logger = LogManager.getLogger(InitializePolicyContextStep.class);
 
-    InitializePolicyContextStep(Step.StepKey key, StepKey nextStepKey) {
+    private final String policy;
+
+    InitializePolicyContextStep(String policy, Step.StepKey key, StepKey nextStepKey) {
         super(key, nextStepKey);
+        this.policy = policy;
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: We don't need to pass the policy in and store it in the step, we can get it directly out of the index metadata (the `index.lifecycle.name` setting)

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java [hunk 8]
```diff
@@ -72,4 +80,22 @@ public ClusterState performAction(Index index, ClusterState clusterState) {
     public boolean isRetryable() {
         return true;
     }
+
+    String policy() {
+        return policy;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (dakrone) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: If we get the policy id from the index metadata we don't need to add this stuff either, which will be nice.

**Review by dakrone — CHANGES_REQUESTED** [review]
I left another comment about not passing the policy in, just deriving it from the index metadata

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java (hunk 7 above)
Comment (andreidan) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/InitializePolicyContextStep.java]: Ah yes, great point!

**Review by dakrone — APPROVED** [review]
LGTM, thanks Andrei!
